[dnf5] General Query and Sack classes

jrohel opened the pull request:
I can prepare more PRs. But they will depended on other code. :(

The PR depend on:
https://github.com/rpm-software-management/libdnf/pull/925
https://github.com/rpm-software-management/libdnf/pull/928
https://github.com/rpm-software-management/libdnf/pull/929

jrohel committed:
[dnf5] Add Query support

jrohel committed:
[dnf5] Add common Sack

j-mracek commented on include/libdnf/common/sack/query.hpp:
```diff
@@ -0,0 +1,243 @@
+        }
+    }
+    return *this;
+}
```
I am sorry, but the search will be probably quite slow.
Iteration over the whole dataset and to compare each element with the search pattern according to the comparison type. Then it removes not matching elements from dataset.
We have to discuss it personally.

j-mracek commented on include/libdnf/common/sack/query_cmp.hpp:
```diff
@@ -0,0 +1,81 @@
+// https://docs.djangoproject.com/en/3.0/ref/models/querysets/#field-lookups
+enum class QueryCmp : uint32_t {
+    // MODIFIERS
+    NOT = (1 << 0),    // negation
```
Believe that we have to create more combination with not.

I am also not sure what is valid for version o release? It is a string but compared like a number.

j-mracek commented on libdnf/common/sack/match_string.cpp:
```diff
@@ -0,0 +1,131 @@
+        case QueryCmp::REGEX: {
+            std::regex re(pattern);
+            result = std::regex_match(value, re);
+        } break;
```
Multiple formats is used here. Please could you unify it? Intead of break return can be used, but I don't know what is preferred.

jrohel replied:
Yes I will unify it.

j-mracek commented on libdnf/common/sack/match_string.cpp:
```diff
@@ -0,0 +1,131 @@
+
+bool match_string(const std::string & value, QueryCmp cmp, const std::string & pattern) {
+    bool result = false;
+    switch (cmp) {
```
The switch has also one disadvantage, that it is tested for each value.

jrohel committed:
[dnf5] Example query implementation based on sack::Query with unit test

jrohel commented:
I made changes:
- ifilter methods do not returns value
- added unit test that implements and tests query based on libdnf::sack::Query with 3 filters
- rebased
